### PR TITLE
Add Partials to Codegen

### PR DESCRIPTION
**Pull request by jamiebrynes7**
#### Description
This adds the partials from the Unity SDK back into the GDK by including them as embedded resources in the code generator. Long term - this will be replaced with a full solution in a modular code generator, however, that work is not in the near term.

As a result, the `Standard Library Utils` feature module has been removed as the functionality was wrapped into the partials 🎉 

#### Tests
Ran code gen! 

#### Documentation
How is this documented (for example: release note, upgrade guide, feature page, in-code documentation)?
Will be in the changelog. Will go through docs and purge mention of the standard library utils if there is one.

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3f [hunk 1]
```diff
@@ -0,0 +1,105 @@
+/// <summary>
+///     A Vector3f with 0f in each dimension.
+/// </summary>
+public static readonly Vector3f ZERO = new Vector3f(0f, 0f, 0f);
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3f]: Not the convention we use elsewhere.

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,12 +8,19 @@
 - Added a reference to Blank Starter Project in the README.
 - Added Frames Per Second (FPS) and Unity heap usage as metrics sent by `MetricSendSystem.cs`.
 - Added a warning message to the top of schema files copied into the `from_gdk_packages` directory.
+- Added the code generation partials for `Coordinates`, `Vector3f`, and `Vector3d` back in from the SDK for Unity.
+    - This adds operators, conversion methods, and utility methods onto these structs.
+    - This supercedes the `StandardLibraryUtils` feature module which was removed as a consequence.
```
Comment (firtina-improbable) [inline CHANGELOG.md]: `supersedes` is more common usage

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3f [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+public Vector3f Normalized()
+{
+    var magnitude = (float) System.Math.Sqrt(SquareMagnitude());
+    return new Vector3f(X / magnitude, Y / magnitude, Z / magnitude);
+}
+
+/// <summary>
+///     True if all components of the Vector3f are real numbers.
+/// </summary>
+public bool IsFinite
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3f]: Why make this a property when everything else is a method?

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/GdkCodeGenerator.csproj [hunk 4]
```diff
@@ -11,6 +11,13 @@
     <Compile Remove="dependencies\**" />
     <EmbeddedResource Remove="dependencies\**" />
     <None Remove="dependencies\**" />
+
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/GdkCodeGenerator.csproj]: Nit: This is the only blank line in the file

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+///     Interpolates between to global coordinates
+/// </summary>
+/// <param name="currentPosition">The starting position</param>
+/// <param name="newPosition">The position to intepolate towards</param>
+/// <param name="progressRatio">
+///     where 0 is the currentPosition and 1 is newPosition. Note: the value is clamped between 0
+///     and 1 to prevent extrapolation.
+/// </param>
+/// <returns>The interpolated position</returns>
+public static Coordinates Lerp(Coordinates currentPosition, Coordinates newPosition, float progressRatio)
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Why are we adding these to coordinates and nothing else?

Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Could make this generic if lots of things are using it.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates [hunk 6]
```diff
@@ -0,0 +1,166 @@
+public static readonly Coordinates ZERO = new Coordinates(0, 0, 0);
+
+public static Vector3d operator -(Coordinates v1, Coordinates v2)
+{
+    return new Vector3d(v1.X - v2.X, v1.Y - v2.Y, v1.Z - v2.Z);
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Should this return a `Coordinates`?

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+{
+    return new Coordinates(v1.X - v2.x, v1.Y - v2.y, v1.Z - v2.z);
+}
+
+public static Coordinates operator +(Coordinates a, Vector3 b)
+{
+    return new Coordinates(a.X + b.x, a.Y + b.y, a.Z + b.z);
+}
+
+public static Coordinates operator -(Vector3d v1, Coordinates v2)
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Should these return `Vector3d` and be part of `Vector3d` extensions?

Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Similar with `f` variant

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+{
+    return new Coordinates(a.x + b.X, a.y + b.Y, a.z + b.Z);
+}
+
+public static bool operator ==(Coordinates a, Coordinates b) => a.Equals(b);
+public static bool operator !=(Coordinates a, Coordinates b) => !a.Equals(b);
+
+public override bool Equals(object obj) 
+{
+    if (obj is Coordinates == false)
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: shortcut: can do 
```c#
if (obj is Coordinates otherCoord)
{
    return X == otherCoord.X && ...
}

return false;
```

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    return new Vector3d(X, Y, Z);
+}
+
+/// <summary>
+///     Returns the string representation of the Coordinates.
+/// </summary>
+public override string ToString()
+{
+    return "Coordinates(" + X + ", " + Y + ", " + Z + ")";
+}
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: `\n` at EOD

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3f [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    return Quaternion.Euler(ToUnityVector());
+}
+
+/// <summary>
+///     Returns the string representation of the Vector3f.
+/// </summary>
+public override string ToString()
+{
+    return "Vector3f(" + X + ", " + Y + ", " + Z + ")";
+}
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3f]: `\n`

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3d [hunk 11]
```diff
@@ -0,0 +1,97 @@
+/// <summary>
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3d]: If we are implementing all of these we should also do IEquatable et al.

File: CHANGELOG.md (hunk 2 above)
Comment (jamiebrynes7) [reply CHANGELOG.md]: Yes you are correct

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 9 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: That will cause some extra new lines when its inserted into the generated file iirc

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3d (hunk 11 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3d]: Agreed - this should be a wider pass over all generated types

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 8 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: 🎉 C# 7 features!

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 6 above)
Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: The idea is that coordinate is an absolute and vector3d is a diff. I personally hate this though.

Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: I think we should probably change this to coordinate - coordinate = coordinate. Once we have custom codegen I would expect a different type to take this job over.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 5 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Discussed offline - not in this PR at the very least

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 8 above)
Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: I generally prefer having logic indented as little as possible.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3f (hunk 3 above)
Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3f]: The standard C# convention is to make this a static method (as primitives can't have instance methods on them).
Vector3 seems a little all over the place so I'm not sure we should try to match that.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates [hunk 12]
```diff
@@ -0,0 +1,166 @@
+public static readonly Coordinates ZERO = new Coordinates(0, 0, 0);
+
+public static Vector3d operator -(Coordinates v1, Coordinates v2)
+{
+    return new Vector3d(v1.X - v2.X, v1.Y - v2.Y, v1.Z - v2.Z);
+}
+
+public static Coordinates operator -(Coordinates v1, Vector3d v2)
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: These mixed Vector/Coordinate operations feel odd. 
Is there some practical use case that makes sense for subtracting a direction from a position, for example?

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+/// <param name="currentPosition">The starting position</param>
+/// <param name="newPosition">The position to intepolate towards</param>
+/// <param name="progressRatio">
+///     where 0 is the currentPosition and 1 is newPosition. Note: the value is clamped between 0
+///     and 1 to prevent extrapolation.
+/// </param>
+/// <returns>The interpolated position</returns>
+public static Coordinates Lerp(Coordinates currentPosition, Coordinates newPosition, float progressRatio)
+{
+    progressRatio = progressRatio > 1 ? 1 : (progressRatio < 0 ? 0 : progressRatio); // Clamp
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Comment usefulness is dubious

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+/// <summary>
+///     Check if a coordinate is near another.
+/// </summary>
+/// <param name="other">The coordinate to test.</param>
+/// <param name="distance">The allowed range.</param>
+/// <returns>True if the other coordinate is within strictly less than the specified range.</returns>
+public bool IsWithinDistance(Coordinates other, double distance)
+{
+#pragma warning disable 618
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Why is this warning disable necessary?

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 9 above)
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Needs newline at end

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 5 above)
Comment (jared-improbable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Could the ratio be better a `double`?

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 12 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: These were all removed in a later commit

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 9 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Will cause weird spacing when its inserted into the code 😕

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 14 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Not sure if it is. This is all swiped from the SDK. This appears to be related to the `Obsolete` attribute 🤔 https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0618

Will remove and see what happens

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 8 above)
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: switch would also work

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Yes it would also work

**Comment by samiwh** [comment]
Ho many of these functions do we actually need? A lot of comments seem to say they are only being added because they were in the SDk.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 12 above)
Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: In the old paradigm of doing things a vector3d is an offset not a direction though.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates (hunk 14 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Removed

File: CHANGELOG.md [hunk 15]
```diff
@@ -8,12 +8,19 @@
 - Added a reference to Blank Starter Project in the README.
 - Added Frames Per Second (FPS) and Unity heap usage as metrics sent by `MetricSendSystem.cs`.
 - Added a warning message to the top of schema files copied into the `from_gdk_packages` directory.
+- Added the code generation partials for `Coordinates`, `Vector3f`, and `Vector3d` back in from the SDK for Unity.
```
[the remark this answers is not in the document]
Comment (jamiebrynes7) [reply CHANGELOG.md]: I've reworded this - have a look 😄

**Comment by jamiebrynes7** [comment]
Stripped down the methods to just operators, explicit conversions between types, and the `Zero` static method

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+public static Coordinates operator /(Coordinates coords, double d)
+{
+    return new Coordinates(coords.X / d, coords.Y / d, coords.Z / d);
+}
+
+public static bool operator ==(Coordinates a, Coordinates b) => a.Equals(b);
+public static bool operator !=(Coordinates a, Coordinates b) => !a.Equals(b);
+
+public override bool Equals(object obj)
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: This is the only one we have defined Equals for. Even if not doing IEquatable in this pass I think we should still have this for the other ones too. Also I have a mild preference for having object overrides together, but not a strong one.

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Coordinates]: Added `GetHashCode` and the equality operators to all the types.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3d [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+        hashCode = (hashCode * 397) ^ Y.GetHashCode();
+        hashCode = (hashCode * 397) ^ Z.GetHashCode();
+        return hashCode;
+    }
+}
+
+/// <summary>
+///     Converts the Vector3d to a Unity Vector3.
+/// </summary>
+public Vector3 ToUnityVector()
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Vector3d]: Mild preference for keeping object overrides together.

**Pull request merged**